pkg/core/bundle_list.go

```diff
@@ -12,52 +13,139 @@ import (
 )
 
 const (
-	maxBundlesToList = 1000000
+	maxMetaFilesToProcess = 1000000
+	typicalBundlesNum     = 1000
 )
 
+func minInt(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// ListBundles returns the list of bundle descriptors from a repo.
+//
+// NOTE: resulting bundles are returned in no particular order.
+//
 // TODO: return a paginated list of id<->bd (separate function in repo_list.go)
-func ListBundles(repo string, store storage.Store) ([]model.BundleDescriptor, error) {
+func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]model.BundleDescriptor, error) {
+	settings := CoreSettings{concurrentBundleList: defaultBundleListConcurrency}
+	for _, bApply := range opts {
+		bApply(&settings)
+	}
+
 	// Get a list
 	e := RepoExists(repo, store)
 	if e != nil {
 		return nil, e
 	}
-	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxBundlesToList)
+
+	// this call is not made async yet
+	ks, _, err := store.KeysPrefix(context.Background(), "", model.GetArchivePathPrefixToBundles(repo), "", maxMetaFilesToProcess)
 	if err != nil {
 		return nil, err
 	}
-	bds := make([]model.BundleDescriptor, 0)
-	for _, k := range ks {
+
+	var (
+		workers, wg sync.WaitGroup
+		werr        error
+	)
+
+	bundleChan := make(chan model.BundleDescriptor)
+	keyChan := make(chan string)
+	errorChan := make(chan error)
```

Comment:
i kinda liked the pattern where both errors and results got sent through the same channel via a struct .. having a separate error channel is okay, too.  was there a reason that the original design was changed from multiplexing both errors and okay results through the same channel to using two separate channels?

Reply:
I kinda liked your suggestion, as I had always made it multiplexed before... :)